### PR TITLE
Bot Action Socket Emit

### PR DESCRIPTION
**Pull request by tausifmuzaffar**
Add the socket emit to Bot Action

### CODE REVIEW

**Comment by tausifmuzaffar** [comment]
@BrandonMurnane let me know if this looks right for socket stuff

**Comment by BrandonMurnane** [comment]
looking good

File: realtime/redisPublisher.js [hunk 1]
```diff
@@ -83,7 +83,7 @@ function publishObject(inst, event, changedKeys, ignoreAttributes) {
   }
 
   if (obj[event]) {
-    if (rtUtils.isRoom(inst)) {
+    if (rtUtils.isRoom(inst) || rtUtils.isBotAction(inst)) {
```
Comment (BrandonMurnane) [inline realtime/redisPublisher.js]: we might wanna split this into a function if we are gonna have more

File: tests/realtime/realtimeUtils.js [hunk 2]
```diff
@@ -148,6 +166,12 @@ describe('tests/realtime/realtimeUtils.js, realtime utils Tests >', () => {
         .to.equal(true);
       });
 
+      it('for should return true for botActionTest', () => {
```
Comment (iamigo) [inline tests/realtime/realtimeUtils.js]: Is there an extra "for" here or am I just not understanding something?

File: realtime/redisPublisher.js [hunk 3]
```diff
@@ -58,6 +58,19 @@ function prepareToPublish(inst, changedKeys, ignoreAttributes) {
 } // prepareToPublish
 
 /**
+ * Determine if an instance is needs to use the bot channel.
```
Comment (iamigo) [inline realtime/redisPublisher.js]: grammar?

File: realtime/redisPublisher.js [hunk 4]
```diff
@@ -58,6 +58,19 @@ function prepareToPublish(inst, changedKeys, ignoreAttributes) {
...
+ *
+ * @param  {Object} inst - Model instance to be published
+ * @returns {Boolean} - True when bot channel is expected
+ */
+function useBotChannel(inst){
+  if (rtUtils.isRoom(inst) || rtUtils.isBotAction(inst)) {
```
Comment (iamigo) [inline realtime/redisPublisher.js]: Seems like jumping through a lot hoops here to check for whether the JSON object is a certain type. What about adding an optional "options" argument to the publishObject function, and just pass in whatever details you need like { isRoom: true, isBotAction: false, ... } -- that way the caller of the publishObject function is responsible for identifying it correctly, which is a lot less ambiguous than letting the receiver try to discern.

**Comment by iamigo** [comment]
Also, looks like you did not run jscs-enforce locally before submitting creating PR. Please *everyone* get in the habit--will save everyone time.

**Comment by tausifmuzaffar** [comment]
@iamigo @shriramshankar @BrandonMurnane I believe i have implemented the changes yall recommended please let me know if there is anything else

**Comment by iamigo** [comment]
Let's talk about modifying the publishObject to include more details so we can avoid stuff like "am I a sample" or "am I a bot" or "am I a subject" or whatever...

File: db/model/botAction.js [hunk 5]
```diff
@@ -25,6 +25,10 @@ const botActionEventNames = {
   upd: 'refocus.internal.realtime.bot.action.update',
   del: 'refocus.internal.realtime.bot.action.remove',
 };
+const pubOpts = {
+  client: 'pubBot',
```
Comment (iamigo) [inline db/model/botAction.js]: what is "pubBot"?

File: realtime/redisPublisher.js [hunk 6]
```diff
@@ -69,9 +69,16 @@ function prepareToPublish(inst, changedKeys, ignoreAttributes) {
  * model that should be ignored
  * @returns {Object} - object that was published
  */
-function publishObject(inst, event, changedKeys, ignoreAttributes) {
+function publishObject(inst, event, changedKeys, ignoreAttributes, opts) {
   const obj = {};
   obj[event] = inst;
+  let pubClient = pubPerspective;
+  let channelName = perspectiveChannelName;
+  if (opts) {
```
Comment (iamigo) [inline realtime/redisPublisher.js]: Add some comments here describing this block?

File: realtime/redisPublisher.js [hunk 7]
```diff
@@ -83,11 +90,7 @@ function publishObject(inst, event, changedKeys, ignoreAttributes) {
   }
 
   if (obj[event]) {
-    if ('useBotChannel' in inst) {
-      pubBot.publish(botChannelName, JSON.stringify(obj));
-    } else {
-      pubPerspective.publish(perspectiveChannelName, JSON.stringify(obj));
-    }
+    pubClient.publish(channelName, JSON.stringify(obj));
```
Comment (iamigo) [inline realtime/redisPublisher.js]: Yay!

**Comment by kamilsmuga** [comment]
@tausifmuzaffar test failures

File: realtime/utils.js [hunk 8]
```diff
@@ -50,6 +50,17 @@ function isThisSubject(obj) {
...
   return obj.hasOwnProperty('type') && obj.hasOwnProperty('settings');
 }
+
+/**
+ * A function to see if an instance is an instance of a room
+ * Checks the name from the model
+ * @param  {Object}  obj - An object instance
+ * @returns {Boolean} - returns true if the name singular is room
+ */
+function isBotAction(obj) {
```
Comment (iamigo) [inline realtime/utils.js]: still?

Comment (tausifmuzaffar) [reply realtime/utils.js]: We need this for the applyFilter in botEmit

Comment (iamigo) [reply realtime/utils.js]: OK, but can you create and prioritize a separate user story to deal with this so we don't have to rely on the presence of attribute names to identify the model?

Comment (tausifmuzaffar) [reply realtime/utils.js]: Will do I created W-4481180 for this

**Comment by tausifmuzaffar** [comment]
@BrandonMurnane @shriramshankar this code is ready for review

**Comment by tausifmuzaffar** [comment]
@shriramshankar @iamigo can we try to close this today? I am available if yall would like to do a live review

**Comment by tausifmuzaffar** [comment]
@pallavi2209 could you review this as shriram is on holiday?

**Pull request merged**